File: scripts/merge-manual-data.js
#!/usr/bin/env node
/**
 * Merge Manual Data — merges Manus AI deep-research output into dashboard-state.json
 *
 * Usage:  node scripts/merge-manual-data.js [path/to/manual-update.json]
 *         Defaults to: data/manual-update.json
 *
 * This script:
 *   1. Reads the manual update file (produced by Manus deep research)
 *   2. Appends new kinetic data points (intercept rate, debris) to existing arrays
 *   3. Updates safety score, briefing, infrastructure statuses
 *   4. Recalculates neighborhood local scores based on new debris data
 *   5. Updates the lastUpdated timestamp
 *   6. Writes the merged result back to dashboard-state.json
 */

const fs = require('fs');
const path = require('path');

const DATA_FILE = path.join(__dirname, '..', 'data', 'dashboard-state.json');
const DEFAULT_MANUAL = path.join(__dirname, '..', 'data', 'manual-update.json');

const manualFile = process.argv[2] || DEFAULT_MANUAL;

// ── Load Files ──────────────────────────────────────────

let state, manual;

try {
  state = JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'));
} catch (e) {
  console.error('❌ Cannot read dashboard-state.json:', e.message);
  process.exit(1);
}

try {
  manual = JSON.parse(fs.readFileSync(manualFile, 'utf8'));
} catch (e) {
  console.error(`❌ Cannot read manual update file (${manualFile}):`, e.message);
  process.exit(1);
}

console.log('═══════════════════════════════════════════════');
console.log('  Merging Manual Research Data');
console.log(`  Source: ${manualFile}`);
console.log(`  Date:   ${manual._date || 'Not specified'}`);
console.log('═══════════════════════════════════════════════\n');

// ── 1. Safety Score ─────────────────────────────────────

if (manual.nationalSafetyScore) {
  const prev = state.nationalSafetyScore.overall;
  state.nationalSafetyScore = manual.nationalSafetyScore;
  // Auto-calculate dayOverDay if not explicitly provided
  if (manual.nationalSafetyScore.dayOverDay === 0 && prev) {
    state.nationalSafetyScore.dayOverDay = manual.nationalSafetyScore.overall - prev;
  }
  console.log(`✅ Safety Score: ${prev} → ${state.nationalSafetyScore.overall} (${state.nationalSafetyScore.dayOverDay >= 0 ? '+' : ''}${state.nationalSafetyScore.dayOverDay})`);
}

// ── 2. Daily Briefing ───────────────────────────────────

if (manual.dailyBriefing) {
  state.dailyBriefing = manual.dailyBriefing;
  console.log(`✅ Daily Briefing updated (${manual.dailyBriefing.length} chars)`);
}

// ── 3. Kinetic Data (append new data points) ────────────

if (manual.kineticData?.national) {
  const kn = manual.kineticData.national;

  // Intercept rate — append to array
  if (kn.interceptRate?.date) {
    const exists = state.kineticData.national.interceptRate.some(d => d.date === kn.interceptRate.date);
    if (!exists) {
      state.kineticData.national.interceptRate.push(kn.interceptRate);
      console.log(`✅ Intercept Rate: added ${kn.interceptRate.date} → ${kn.interceptRate.value}%`);
    } else {
      console.log(`⏭️  Intercept Rate: ${kn.interceptRate.date} already exists, skipping`);
    }
    // Also add to dateLabels if not present
    if (!state.kineticData.dateLabels.includes(kn.interceptRate.date)) {
      state.kineticData.dateLabels.push(kn.interceptRate.date);
    }
  }

  // Debris incidents — append to array
  if (kn.debrisIncidents?.date) {
    const exists = state.kineticData.national.debrisIncidents.some(d => d.date === kn.debrisIncidents.date);
    if (!exists) {
      state.kineticData.national.debrisIncidents.push(kn.debrisIncidents);
      console.log(`✅ Debris Incidents (national): added ${kn.debrisIncidents.date} → ${kn.debrisIncidents.value}`);
    } else {
      console.log(`⏭️  Debris Incidents: ${kn.debrisIncidents.date} already exists, skipping`);
    }
  }
}

// ── 4. Neighborhood Debris (append to per-neighborhood arrays) ──

if (manual.kineticData?.byNeighborhood) {
  const byN = manual.kineticData.byNeighborhood;
  let updated = 0;

  for (const [name, count] of Object.entries(byN)) {
    if (name.startsWith('_')) continue; // skip _note fields
    if (state.kineticData.byNeighborhood[name]) {
      state.kineticData.byNeighborhood[name].debrisIncidents.push(count);

      // Recalculate local score based on recent debris:
      // Base national score minus penalty for recent debris
      const recentDebris = state.kineticData.byNeighborhood[name].debrisIncidents.slice(-5);
      const debrisSum = recentDebris.reduce((a, b) => a + b, 0);
      const nationalScore = state.nationalSafetyScore.overall;
      let localScore = Math.max(0, Math.min(100, nationalScore - (debrisSum * 4)));
      state.kineticData.byNeighborhood[name].localScore = localScore;

      // Update status tier
      if (localScore >= 75) state.kineticData.byNeighborhood[name].status = 'safe';
      else if (localScore >= 50) state.kineticData.byNeighborhood[name].status = 'caution';
      else if (localScore >= 25) state.kineticData.byNeighborhood[name].status = 'warning';
      else state.kineticData.byNeighborhood[name].status = 'critical';

      updated++;
    }
  }
  console.log(`✅ Neighborhood debris: updated ${updated} areas`);
}

// ── 5. Infrastructure Statuses ──────────────────────────

if (manual.infrastructure) {
  for (const [key, val] of Object.entries(manual.infrastructure)) {
    if (state.infrastructure[key]) {
      state.infrastructure[key] = val;
    }
  }
  console.log('✅ Infrastructure statuses updated');
}

// ── 6. Update Timestamp ─────────────────────────────────

state.lastUpdated = new Date().toISOString();
console.log(`\n✅ lastUpdated → ${state.lastUpdated}`);

// ── Write Back ──────────────────────────────────────────

fs.writeFileSync(DATA_FILE, JSON.stringify(state, null, 2), 'utf8');
console.log(`\n💾 Written to ${DATA_FILE}`);
console.log('═══════════════════════════════════════════════\n');

File: scripts/merge-manual-data.py
#!/usr/bin/env python3
import sys
import os
import json
import datetime

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DATA_FILE = os.path.join(BASE_DIR, 'data', 'dashboard-state.json')
DEFAULT_MANUAL = os.path.join(BASE_DIR, 'data', 'manual-update.json')

manual_file = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_MANUAL

try:
    with open(DATA_FILE, 'r', encoding='utf-8') as f:
        state = json.load(f)
except Exception as e:
    print(f"❌ Cannot read dashboard-state.json: {e}")
    sys.exit(1)

try:
    with open(manual_file, 'r', encoding='utf-8') as f:
        manual = json.load(f)
except Exception as e:
    print(f"❌ Cannot read manual update file ({manual_file}): {e}")
    sys.exit(1)

print('═══════════════════════════════════════════════')
print('  Merging Manual Research Data')
print(f"  Source: {manual_file}")
print(f"  Date:   {manual.get('_date', 'Not specified')}")
print('═══════════════════════════════════════════════\n')

# 1. Safety Score
if "nationalSafetyScore" in manual:
    prev = state.get("nationalSafetyScore", {}).get("overall", 0)
    state["nationalSafetyScore"] = manual["nationalSafetyScore"]
    
    if manual["nationalSafetyScore"].get("dayOverDay") == 0 and prev:
        state["nationalSafetyScore"]["dayOverDay"] = manual["nationalSafetyScore"]["overall"] - prev
        
    overall = state["nationalSafetyScore"]["overall"]
    dod = state["nationalSafetyScore"].get("dayOverDay", 0)
    sign = "+" if dod >= 0 else ""
    print(f"✅ Safety Score: {prev} → {overall} ({sign}{dod})")

# 2. Daily Briefing
if "dailyBriefing" in manual:
    state["dailyBriefing"] = manual["dailyBriefing"]
    print(f"✅ Daily Briefing updated ({len(manual['dailyBriefing'])} chars)")

# 3. Kinetic Data
kinetic_manual = manual.get("kineticData", {})
kinetic_state = state.setdefault("kineticData", {})
kinetic_state_national = kinetic_state.setdefault("national", {})

if "national" in kinetic_manual:
    kn = kinetic_manual["national"]
    
    # Intercept rate
    ir = kn.get("interceptRate")
    if ir and "date" in ir:
        ir_list = kinetic_state_national.setdefault("interceptRate", [])
        exists = any(d.get("date") == ir["date"] for d in ir_list)
        if not exists:
            ir_list.append(ir)
            print(f"✅ Intercept Rate: added {ir['date']} → {ir['value']}%")
        else:
            print(f"⏭️  Intercept Rate: {ir['date']} already exists, skipping")
            
        date_labels = kinetic_state.setdefault("dateLabels", [])
        if ir["date"] not in date_labels:
            date_labels.append(ir["date"])
            
    # Debris
    db = kn.get("debrisIncidents")
    if db and "date" in db:
        db_list = kinetic_state_national.setdefault("debrisIncidents", [])
        exists = any(d.get("date") == db["date"] for d in db_list)
        if not exists:
            db_list.append(db)
            print(f"✅ Debris Incidents (national): added {db['date']} → {db['value']}")
        else:
            print(f"⏭️  Debris Incidents: {db['date']} already exists, skipping")

# 4. Neighborhood Debris
if "byNeighborhood" in kinetic_manual:
    byn = kinetic_manual["byNeighborhood"]
    state_byn = kinetic_state.setdefault("byNeighborhood", {})
    updated = 0
    for name, count in byn.items():
        if name.startswith('_'):
            continue
        if name in state_byn:
            # Append local debris count
            state_byn[name].setdefault("debrisIncidents", []).append(count)
            
            recent_debris = state_byn[name]["debrisIncidents"][-5:]
            debris_sum = sum(recent_debris)
            national_score = state.get("nationalSafetyScore", {}).get("overall", 0)
            
            local_score = max(0, min(100, national_score - (debris_sum * 4)))
            state_byn[name]["localScore"] = local_score
            
            if local_score >= 75:
                status = 'safe'
            elif local_score >= 50:
                status = 'caution'
            elif local_score >= 25:
                status = 'warning'
            else:
                status = 'critical'
                
            state_byn[name]["status"] = status
            updated += 1
            
    print(f"✅ Neighborhood debris: updated {updated} areas")

# 5. Infrastructure
if "infrastructure" in manual:
    state_infra = state.setdefault("infrastructure", {})
    for key, val in manual["infrastructure"].items():
        if key in state_infra:
            state_infra[key] = val
    print('✅ Infrastructure statuses updated')

# 6. Update Timestamp
state["lastUpdated"] = datetime.datetime.now(datetime.timezone.utc).isoformat().replace("+00:00", "Z")
print(f"\n✅ lastUpdated → {state['lastUpdated']}")

with open(DATA_FILE, 'w', encoding='utf-8') as f:
    json.dump(state, f, indent=2)

print(f"\n💾 Written to {DATA_FILE}")
print('═══════════════════════════════════════════════\n')
